### DISCUSSION TITLE
@SpanAttribute annotation with custom value resolver for an attribute

### DISCUSSION THREAD

**Discussion by ashwin-07** [open-telemetry/opentelemetry-java-instrumentation#7484] [Q&A] [answered]
Hello 👋

I was wondering if there was any value resolver kind of thing which is present for @SpanAttribute annotation for auto instrumentation. So that we could pass in something like `@SpanAttribute(resolver=com.test.customAttrResolver.class)` and the return value of the custom resolver would be recorded as value for the span attribute

### Example:
Adding a sample code for illustration.

_**TestObject.java**_
```
public class TestObject {
    String field1;   //only this field should be recorded as value for the span attribute
    String field2;
    String field3;
}
```

_**Instrumented Method**_
```

@WithSpan("test method")
 private void processChange(@SpanAttribute TestObject testObject) {
    /*business logic*/
 }
```

I am looking for an option to record only the testObject.field1 property and leave out the rest two properties of the Span Attribute testObject as part of the span. 

I guess overriding the toString method will help achieve this but I cannot override toString in my actual codebase because it is also used in other places as well. Is there any way to achieve this with SpanAttribute?

**Comment by mateuszrzeszutek** [comment, marked as answer]
Hey @ashwin-07 ,

Have you tried using `Span.current()` and just setting the value manually?
```java
@WithSpan("test method")
private void processChange(TestObject testObject) {
    Span.current().setAttribute("my-custom-attr", testObject.field1);
    /*business logic*/
}
```
I believe it's a bit easier than implementing a custom resolver class (and takes less code).

**Reply by ashwin-07** [reply]
Yeah I am currently doing it that way, and it makes sense. But, We were previously using Sleuth and there was a [custom extractor](https://cloud.spring.io/spring-cloud-sleuth/reference/html/#custom-extractor) option for the equivalent SpanTag annotation. So was just curious to know if there is a way to do the same with the SpanAttribute

**Reply by mateuszrzeszutek** [reply]
> So was just curious to know if there is a way to do the same with the SpanAttribute

No, there is not. In cases like this one, we usually recommend using `Span.current()`.

**Reply by ashwin-07** [reply]
oh okay, thanks!

**Reply by trask** [reply]
(just fyi, there is one downside to using Span.current().setAttribute() which is that those attributes aren't available for attribute rule-based sampling, though this is generally a more advanced use-case)

**Reply by ashwin-07** [reply]
@trask I am using a custom sampler to filter out traces based on the span name, don't have a use case for sampling based on attributes. Thanks for pointing it out anyways
